chalk-solve/src/infer/instantiate.rs

```diff
@@ -95,28 +93,45 @@ impl InferenceTable {
                 }
             })
             .collect();
-        Subst::apply(&parameters, value)
+        Subst::apply(&parameters, &value)
     }
 }
 
-pub(crate) trait BindersAndValue {
-    type Output;
+pub(crate) trait IntoBindersAndValue {
+    type Binders: IntoIterator<Item = ParameterKind<()>>;
+    type Value;
 
-    fn split(&self) -> (&[ParameterKind<()>], &Self::Output);
+    fn into_binders_and_value(self) -> (Self::Binders, Self::Value);
 }
 
-impl<T> BindersAndValue for Binders<T> {
-    type Output = T;
+impl<'a, T> IntoBindersAndValue for &'a Binders<T> {
+    type Binders = std::iter::Cloned<std::slice::Iter<'a, ParameterKind<()>>>;
+    type Value = &'a T;
+
+    fn into_binders_and_value(self) -> (Self::Binders, Self::Value) {
+        (self.binders.iter().cloned(), &self.value)
+    }
+}
+
+impl<'a> IntoBindersAndValue for &'a QuantifiedTy<ChalkIr> {
+    type Binders = std::iter::Map<std::ops::Range<usize>, fn(usize) -> chalk_ir::ParameterKind<()>>;
+    type Value = &'a Ty<ChalkIr>;
+
+    fn into_binders_and_value(self) -> (Self::Binders, Self::Value) {
+        fn make_lifetime(_: usize) -> ParameterKind<()> {
+            ParameterKind::Lifetime(())
+        }
 
-    fn split(&self) -> (&[ParameterKind<()>], &Self::Output) {
-        (&self.binders, &self.value)
+        let p: fn(usize) -> ParameterKind<()> = make_lifetime;
+        ((0..self.num_binders).map(p), &self.ty)
```

Comment:
Because you can't name the resulting type -- the type of `make_lifetime` is a unique ZST tied to that function, and it can't be typed. Really what I want here is `impl Trait` but that's not available in this context on stable.